src/lsm/grid.zig

```diff
@@ -337,6 +340,17 @@ pub fn GridType(comptime Storage: type) type {
             assert(!grid.read_recursion_guard);
             grid.assert_not_writing(read.address, null);
 
+            if (grid.superblock.free_set.is_free(read.address)) {
+                // We cannot assert `free_set.is_free()` because of the following case:
+                // 1. The replica receives a request_block from a repairing replica.
+                //    The block is allocated but not cached — but is due to be freed at checkpoint.
+                // 2. All of the Grid's Read IOPS are occupied, so queue the read.
+                // 3. The replica checkpoints.
+                // 4. The read dequeues, but the requested block is no longer allocated.
+                // TODO(State Transfer) How does a non-repairable block trigger state transfer?
```

Comment:
I'm glad that is how it will work! I actually started to make this exact change before, but backed it out because I figured there was a different plan that matched the existing signature :grin:. I will update this TODO, but defer making the "block → optional-block" callback refactor until I implement state transfer.